src/code/z_eventmgr.c

```diff
@@ -1,52 +1,549 @@
+    sActorCsMgr.endCsId = CS_ID_NONE;
+    sActorCsMgr.play = play;
+    sActorCsMgr.length = -1;
+    sActorCsMgr.targetActor = NULL;
+    sActorCsMgr.subCamId = SUB_CAM_ID_DONE;
+    sActorCsMgr.isCameraStored = false;
+    sActorCsMgr.csId = sActorCsMgr.endCsId;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_GetCurrentSubCamId.s")
+/**
+ * Store camera into subCam 2, and keep subCam 2 INACTIVE to preserve the struct.
+ */
+void ActorCutscene_StoreCamera(Camera* camera) {
+    if (camera != NULL) {
+        memcpy(&sActorCsMgr.play->subCameras[2], camera, sizeof(Camera));
+        sActorCsMgr.play->subCameras[2].camId = camera->camId;
+        Camera_ChangeStatus(&sActorCsMgr.play->subCameras[2], CAM_STATUS_INACTIVE);
+        sActorCsMgr.isCameraStored = true;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/func_800F21CC.s")
+void ActorCutscene_ClearWaiting(void) {
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/func_800F22C4.s")
+    for (i = 0; i < ARRAY_COUNT(sWaitingActorCutsceneList); i++) {
+        sWaitingActorCutsceneList[i] = 0;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_eventmgr/ActorCutscene_SetReturnCamera.s")
+void ActorCutscene_ClearNextCutscenes(void) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(sNextActorCutsceneList); i++) {
+        sNextActorCutsceneList[i] = 0;
+    }
+}
+
+s16 ActorCutscene_MarkNextCutscenes(void) {
+    s16 bit;
+    s32 i;
+    s32 j;
+    s32 count = 0;
+    s16 actorCsIdMax = CS_ID_NONE;
+    s16 priorityMax = SHT_MAX;
+    s16 csId;
+    s16 priority;
+
+    for (i = 0; i < ARRAY_COUNT(sNextActorCutsceneList); i++) {
+        for (bit = 1, j = 0; j < 8; j++) {
+            if (sWaitingActorCutsceneList[i] & bit) {
+                csId = (i << 3) | j;
+                priority = ActorCutscene_GetCutsceneImpl(csId)->priority;
+
+                if ((priority ^ 0) == -1) {
+                    sNextActorCutsceneList[i] |= bit;
+                } else if ((priority < priorityMax) && (priority > 0)) {
+                    actorCsIdMax = csId;
+                    priorityMax = priority;
+                }
+                count++;
+            }
+            bit <<= 1;
+        }
+    }
+    if (actorCsIdMax != CS_ID_NONE) {
+        sNextActorCutsceneList[actorCsIdMax >> 3] |= 1 << (actorCsIdMax & 7);
+    }
+    return count;
+}
+
+void ActorCutscene_End(void) {
+    ActorCutscene* actorCs;
+    s16 oldCamId;
+    s16 oldStateFlags;
+
+    switch (sActorCsMgr.startMethod) {
+        case CS_START_2:
+            sActorCsMgr.targetActor->flags &= ~ACTOR_FLAG_100000;
+            // fallthrough
+        case CS_START_1:
+            func_800B7298(sActorCsMgr.play, 0, PLAYER_CSMODE_END);
+            sActorCsMgr.startMethod = CS_START_0;
+            break;
+
+        default:
+            break;
+    }
+
+    actorCs = ActorCutscene_GetCutsceneImpl(sActorCsMgr.csId);
+
+    switch (actorCs->endSfx) {
```

Comment:
This switch still gives me existential dread, why would they design it this way rather than
```c
if (actorCs->endSfx != 0xFFFF)
    play_sound(actorCs->endSfx);
```
(sorry bit tangential, there's nothing we can do about how they wrote it)